### PR TITLE
ci(gpu): gate GPU tests in-workflow so the required check always reports

### PR DESCRIPTION
**Pull request by vivekkhandelwal1**
## The problem

`pytest (GPU, MI350)` is a required status check in the active `main protection` ruleset (`strict_required_status_checks_policy: true`). But `gpu-tests.yml` filtered its `pull_request` trigger with a `paths:` list.

Per [GitHub's docs on required status checks](https://docs.github.com/en/pull-requests/how-tos/merge-and-close-pull-requests/troubleshooting-required-status-checks), those two settings are incompatible:

| Cause | Result |
|---|---|
| A workflow is skipped by path filtering | Associated checks stay **Pending** and block merging |
| A job is skipped by a conditional | The job reports **Success** |

So a PR touching none of the filtered paths never produces a `pytest (GPU, MI350)` result at all, and the merge gate waits forever on a check that will never arrive. This is not theoretical — five of the seven currently open PRs are blocked by it:

| PR | Files | GPU checks |
|---|---|---|
| #307 | `docs/**` | none |
| #313 | `.github/workflows/{lock-requirements,refresh-baselines}.yml` | none |
| #314 | `.github/workflows/*` (setup-python bump) | none |
| #315 | `nightly-eval.yml`, `pages.yml`, `docs/` | none |
| #317 | `src/aorta/cli/**`, `src/aorta/tools/**`, `pyproject.toml` | none |
| #318 | `tests/instrumentation/test_environment.py` | 2 |
| #320 | `scripts/ci/gen_dashboard.py` | 2 |

Note the workflow's own header comment anticipated making this check required, but marking it so didn't account for the paths filter.

## The fix

Move the path decision from the trigger into the workflow:

- `pull_request:` no longer carries `paths:`, so the workflow always starts and always reports.
- A new `changes` job on `ubuntu-latest` lists the PR's files via `gh api .../pulls/N/files` and sets a `gpu` output. Non-PR events (schedule, `workflow_dispatch`) always return `true`.
- `gpu-tests` gains `needs: changes` and gates on `needs.changes.outputs.gpu == 'true'`, keeping its existing fork-PR guard. `gpu-regression` skips along with it, since a dependency that did not succeed skips the dependent job.

A skipped job reports Success, so docs-only PRs satisfy the required check in seconds without ever touching the MI350 runner. `permissions:` gains `pull-requests: read` for the file listing.

**Job names are unchanged**, so the ruleset needs no edit.

## Behaviour is preserved

The pattern list is one-to-one with the `paths:` filter it replaces. I replayed the gate logic against the actual file lists of all seven open PRs and it reproduces the old decisions exactly — `gpu=true` for #318 and #320, `gpu=false` for the other five.

Edge cases also checked: a brand-new file under `scripts/ci/`, a deep subtree under `src/aorta/workloads/`, `gpu-tests.yml` itself, an unrelated workflow, docs-only, and near-miss lookalikes (`scripts/other/x.py`, `docker.md`, `src/aorta/racecar/x.py`) all classify correctly.

One implementation note worth flagging for review: the patterns are single-quoted in a bash array and iterated as `"${patterns[@]}"`. An earlier draft used an unquoted string list, which the shell filename-expanded against the working directory — that silently turned `scripts/ci/*` into whatever happened to exist locally and would have missed newly added files. The array form keeps them literal so they reach `case` as patterns.

## Verification

This PR touches `.github/workflows/gpu-tests.yml`, which is in the pattern list, so GPU jobs should run here and prove the positive path still works. The skip path is proven by updating #315 with `main` afterwards and confirming `pytest (GPU, MI350)` reports as skipped/Success.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: .github/workflows/gpu-tests.yml [hunk 1]
```diff
@@ -45,15 +38,86 @@ env:
...
+              esac
+            done
+          done < <(gh api "repos/${GITHUB_REPOSITORY}/pulls/${PR_NUMBER}/files" \
+                     --paginate --jq '.[].filename')
+
+          if [ "$gpu" = false ]; then
+            echo "No GPU-relevant paths changed; the GPU jobs will be skipped."
+            echo "A skipped job reports Success, satisfying the required check."
+          fi
+          echo "gpu=$gpu" >> "$GITHUB_OUTPUT"
```
[the remark this answers is not in the document]
Comment (vivekkhandelwal1) [reply .github/workflows/gpu-tests.yml:124]: Right conclusion, and fixed in `504d1a3` — but the failure mode was the opposite of the one described, and worse, so worth recording why.

The gate read the listing through process substitution:

```bash
done < <(gh api ".../pulls/${PR_NUMBER}/files" --paginate --jq '.[].filename')
```

A process substitution's exit status is not part of the enclosing command, and `pipefail` only applies to pipelines. So `set -euo pipefail` never saw a failed `gh api`. The job did **not** fail — it read an empty stream, concluded `gpu=false`, and exited 0.

That means the required check was not left unresolved. It went **green** — `gpu-tests` skipped, `pytest (GPU, MI350)` reported Success — on a PR that may well have touched GPU code. A silent false pass on a merge gate rather than a stuck PR.

Confirmed by driving both versions with a stubbed `gh` that exits 1:

| Scenario | Before | After |
|---|---|---|
| `gh api` fails (503 / rate limit / scope) | exit 0, `gpu=false` → **green without testing** | `gpu=true`, warning annotation |
| Listing comes back empty | exit 0, `gpu=false` | `gpu=true`, warning annotation |
| GPU-relevant file present | `gpu=true` | `gpu=true` |
| Docs only | `gpu=false` | `gpu=false` |

The fix captures the output explicitly and defaults to `gpu=true` when it cannot be retrieved or is empty:

```bash
if ! files="$(gh api ".../files" --paginate --jq '.[].filename' 2>&1)"; then
  echo "::warning::Could not list the PR's files, so running the GPU suite..."
  echo "gpu=true" >> "$GITHUB_OUTPUT"
  exit 0
fi
```

Assigning inside `if !` keeps `set -e` from aborting, so the failure is handled rather than swallowed. Iteration is now `done <<< "$files"`. The empty-list guard is there because a real PR always reports at least one file, so an empty listing means something went wrong rather than "nothing to test".

Worth stating the residual limit: `GET /pulls/{n}/files` caps at 3000 files, so a PR larger than that could in principle miss a GPU-relevant path and under-trigger. That exposure is identical to the `paths:` filter this replaced and a PR that size is implausible here, so I have left it rather than adding paging gymnastics — happy to guard it if you would rather.
